### PR TITLE
Fix emit with this and rest parameters

### LINKED ISSUE #9682: Returning a function with "this" typing and rest parameter down-emit incorrect
**TypeScript Version:**  2.0.0 (beta)

**Code**

```typescript
function rebase(fn: (base: any, ...args: any[]) => any): (...args: any[]) => any {
	return function(this: any, ...args: any[]) {
		return fn.apply(this, [ this ].concat(args));
	};
}
```

**Expected behavior:**

For the index to copy arguments to start at 0:

```js
function rebase(fn) {
    return function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        return fn.apply(this, [this].concat(args));
    };
}
```

**Actual behavior:** 

Incorrectly, the index for starting to copy the arguments has been set to 1:

```js
function rebase(fn) {
    return function () {
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        return fn.apply(this, [this].concat(args));
    };
}
```

### PR DESCRIPTION
**Pull request by sandersn**
Fixes #9682 

and also fixes the same problem when emitting decorator metadata.

### CODE REVIEW

File: src/compiler/emitter.ts [hunk 1]
```diff
@@ -6155,10 +6156,11 @@ const _super = (function (geti, seti) {
 
                     if (valueDeclaration) {
                         const parameters = valueDeclaration.parameters;
+                        const skipThisCount = parameters.length && (<Identifier>parameters[0].name).text === "this" ? 1 : 0;
```
Comment (mhegazy) [inline src/compiler/emitter.ts]: `parameters[0].originalKeywordKind === SyntaxKind.ThisKeyword`

**Comment by sandersn** [comment]
@mhegazy are you OK with merging this now?

**Comment by mhegazy** [comment]
👍

**Comment by mhegazy** [comment]
We will need to merge this in release-2.0 as well.

**Comment by sandersn** [comment]
Done. I cherry-picked the commits into release-2.0

**Pull request merged**